File: backend/sdks/javascript/src/exceptions.ts
/**
 * AI-HR Platform SDK Exceptions
 * 
 * Custom error classes for the AI-HR Platform SDK.
 */

/**
 * Base exception for AI-HR Platform SDK
 */
export class AIHRException extends Error {
  public readonly statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = 'AIHRException';
    this.statusCode = statusCode;
    
    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AIHRException);
    }
  }
}

/**
 * Authentication failed
 */
export class AuthenticationError extends AIHRException {
  constructor(message: string = 'Authentication failed') {
    super(message, 401);
    this.name = 'AuthenticationError';
  }
}

/**
 * Authorization failed
 */
export class AuthorizationError extends AIHRException {
  constructor(message: string = 'Insufficient permissions') {
    super(message, 403);
    this.name = 'AuthorizationError';
  }
}

/**
 * Resource not found
 */
export class NotFoundError extends AIHRException {
  constructor(message: string = 'Resource not found') {
    super(message, 404);
    this.name = 'NotFoundError';
  }
}

/**
 * Request validation failed
 */
export class ValidationError extends AIHRException {
  public readonly errors?: Array<Record<string, any>>;

  constructor(message: string = 'Validation failed', errors?: Array<Record<string, any>>) {
    super(message, 422);
    this.name = 'ValidationError';
    this.errors = errors;
  }
}

/**
 * Rate limit exceeded
 */
export class RateLimitError extends AIHRException {
  public readonly retryAfter: number;

  constructor(message: string = 'Rate limit exceeded', retryAfter: number = 60) {
    super(message, 429);
    this.name = 'RateLimitError';
    this.retryAfter = retryAfter;
  }
}

/**
 * Server error
 */
export class ServerError extends AIHRException {
  constructor(message: string = 'Internal server error') {
    super(message, 500);
    this.name = 'ServerError';
  }
}

/**
 * Request timeout
 */
export class TimeoutError extends AIHRException {
  constructor(message: string = 'Request timeout') {
    super(message);
    this.name = 'TimeoutError';
  }
}

/**
 * Connection error
 */
export class ConnectionError extends AIHRException {
  constructor(message: string = 'Connection error') {
    super(message);
    this.name = 'ConnectionError';
  }
}

File: backend/sdks/python/aihr_platform_sdk/exceptions.py
"""
AI-HR Platform SDK Exceptions

Custom exceptions for the AI-HR Platform SDK.
"""

from typing import Optional, Dict, Any, List


class AIHRException(Exception):
    """Base exception for AI-HR Platform SDK"""
    
    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code


class AuthenticationError(AIHRException):
    """Authentication failed"""
    
    def __init__(self, message: str = "Authentication failed"):
        super().__init__(message, 401)


class AuthorizationError(AIHRException):
    """Authorization failed"""
    
    def __init__(self, message: str = "Insufficient permissions"):
        super().__init__(message, 403)


class NotFoundError(AIHRException):
    """Resource not found"""
    
    def __init__(self, message: str = "Resource not found"):
        super().__init__(message, 404)


class ValidationError(AIHRException):
    """Request validation failed"""
    
    def __init__(self, message: str = "Validation failed", errors: Optional[List[Dict[str, Any]]] = None):
        super().__init__(message, 422)
        self.errors = errors or []


class RateLimitError(AIHRException):
    """Rate limit exceeded"""
    
    def __init__(self, message: str = "Rate limit exceeded", retry_after: int = 60):
        super().__init__(message, 429)
        self.retry_after = retry_after


class ServerError(AIHRException):
    """Server error"""
    
    def __init__(self, message: str = "Internal server error"):
        super().__init__(message, 500)


class TimeoutError(AIHRException):
    """Request timeout"""
    
    def __init__(self, message: str = "Request timeout"):
        super().__init__(message)


class ConnectionError(AIHRException):
    """Connection error"""
    
    def __init__(self, message: str = "Connection error"):
        super().__init__(message)